File: 2021/10/17/main.go

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, targetSum int) int {
	if root == nil {
		return 0
	}

	cnt := 0

	var dfs func(*TreeNode, int)
	dfs = func(node *TreeNode, sum int) {
		if node == nil {
			return
		}

		sum -= node.Val
		if sum == 0 {
			cnt++
		}

		dfs(node.Left, sum)
		dfs(node.Right, sum)
	}

	dfs(root, targetSum)

	return cnt + pathSum(root.Left, targetSum) + pathSum(root.Right, targetSum)
}

File: 2021/10/17/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        if root == None:
            return 0
        self.cnt = 0
        def dfs(node, sum):
            if node == None:
                return

            sum -= node.val
            if sum == 0:
                self.cnt += 1

            dfs(node.left, sum)
            dfs(node.right, sum)

        dfs(root, targetSum)

        return self.cnt + self.pathSum(root.left, targetSum) + self.pathSum(root.right, targetSum)